### PR TITLE
Use O(1) additions to the BufferState class.

### PR DESCRIPTION
**Pull request by SandyChapman**
Measurements with capacity:

Value in header represent capacity.
Value in table represents execution time in seconds.

Note that this solution trades memory space for speed. If the code isn't desired, then I suggest at least we add the special case for a buffer containing a single value (since this applies to MutableProperties).

Add Value @ 500,000 iterations | 1 | 10 | 100 | 1000 | 10000 
------------- | ------------- | ------------- | ------------- | ------------- | -------------
now | 1.34 | 1.27 | 1.37 | 3.14 |20.8
new|0.354|0.59|0.587|0.603|0.615
        
        
iterate empty @ 100,000 iterations | 1 | 10 | 100 | 1000 | 10000
------------- | ------------- | ------------- | ------------- | ------------- | -------------
now | 0.0995 | 0.104 | 0.103 | 0.104 | 
new | 0.16 | 0.159 | 0.162 | 0.171 | 
        
        
iterate half full @ 100,000 iterations | 1 | 10 | 100 | 1000 | 10000
------------- | ------------- | ------------- | ------------- | ------------- | -------------
now | 0.167 | 0.444 | 2.96 | 28.9 | 
new | 0.159 | 0.502 | 3.03 | 29 | 
        
        
iterate full @ 100,000 iterations | 1 | 10 | 100 | 1000 | 10000
------------- | ------------- | ------------- | ------------- | ------------- | -------------
now | 0.167 | 0.675 | 5.71 | 57.6 | 
new | 0.156 | 0.722 | 5.95 | 57.7 |

### CODE REVIEW

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 1]
```diff
@@ -275,16 +285,29 @@ private struct BufferState<Value, Error: ErrorType> {
...
-		values.append(value)
-
-		let overflow = values.count - capacity
-		if overflow > 0 {
-			values.removeRange(0..<overflow)
-		}
-	}
+    mutating func addValue(value: Value) {
+        if self.capacity == 1 {
+            self.values = [value]
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Fast path :+1:

**Comment by NachoSoto** [comment]
This is awesome. I've experienced buffer being really slow, so I'm :+1: on this!

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 2]
```diff
@@ -275,16 +285,29 @@ private struct BufferState<Value, Error: ErrorType> {
...
+        values[index] = value
+        values[index + capacity] = value
+        index = (index + 1) % (self.capacity)
+        if count < self.capacity {
+            count = count + 1
+        }
+        self.slice = nil
+    }
+    
+    mutating func generator() -> IndexingGenerator<Array<Value>> {
```
Comment (bencochran) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Could this be `generate()` and `BufferState` be `SequenceType`? Then the `forEach` loop in `replay` above could just be on `originalState` and avoid a force-unwrap.

(Code reviewing on my phone, so it's possible I’m missing something… pre-apologies if so)

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 3]
```diff
@@ -263,9 +265,17 @@ public struct SignalProducer<Value, Error: ErrorType> {
 }
 
 private struct BufferState<Value, Error: ErrorType> {
-	// All values in the buffer.
-	var values: [Value] = []
-
+    
+    init(capacity: Int) {
```
Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Need to convert these spaces to tabs

File: ReactiveCocoa/Swift/SignalProducer.swift (hunk 2 above)
Comment (SandyChapman) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Likely. I wasn't positive on whether I'd need to implement other methods to conform to the SequenceType protocol. I'll look into this though.

File: ReactiveCocoa/Swift/SignalProducer.swift (hunk 1 above)
Comment (SandyChapman) [inline ReactiveCocoa/Swift/SignalProducer.swift]: This is especially good for MutableProperties which always have a capacity of 1.

File: ReactiveCocoa/Swift/SignalProducer.swift (hunk 3 above)
Comment (SandyChapman) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Sure, I'll convert them.

**Comment by RuiAAPeres** [comment]
Ping @SandyChapman

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 4]
```diff
@@ -275,16 +285,29 @@ private struct BufferState<Value, Error: ErrorType> {
...
-	}
+    mutating func addValue(value: Value) {
+        if self.capacity == 1 {
+            self.values = [value]
+            return
+        }
+        if values.count == 0 {
+            values = Array<Value>(count: self.capacity * 2, repeatedValue: value)
+        }
+        values[index] = value
```
Comment (ikesyo) [inline ReactiveCocoa/Swift/SignalProducer.swift]: The test suite crashes here (on my local environment with the latest `master`) in the test case: https://github.com/ReactiveCocoa/ReactiveCocoa/blob/17bd9528eb436cab5d2f388edd17b0e6982f46a8/ReactiveCocoaTests/Swift/SignalProducerLiftingSpec.swift#L1005-L1017

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 5]
```diff
@@ -275,16 +285,29 @@ private struct BufferState<Value, Error: ErrorType> {
...
-			values.removeRange(0..<overflow)
-		}
-	}
+    mutating func addValue(value: Value) {
+        if self.capacity == 1 {
+            self.values = [value]
+            return
+        }
+        if values.count == 0 {
+            values = Array<Value>(count: self.capacity * 2, repeatedValue: value)
```
Comment (ikesyo) [inline ReactiveCocoa/Swift/SignalProducer.swift]: The test suites crashes here (on my local environment with the latest `master`) in the test case: https://github.com/ReactiveCocoa/ReactiveCocoa/blob/17bd9528eb436cab5d2f388edd17b0e6982f46a8/ReactiveCocoaTests/Swift/SignalProducerSpec.swift#L251-L284

Comment (ikesyo) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Passing `Int.max` as the `capacity` easily let this crash.

**Comment by mdiep** [comment]
I'm going to close this PR because it's stale and `buffer` is going away in RAC 5 (#2831).

**Comment by SandyChapman** [comment]
@mdiep : Likely a good idea. I've been meaning to get back to this, but in all reality, I wouldn't be able to find the time.

**Pull request closed without merging**